pkg/query/querier.go

```diff
@@ -209,7 +209,11 @@ func (q *querier) Select(_ bool, hints *storage.SelectHints, ms ...*labels.Match
 	// The querier has a context but it gets canceled, as soon as query evaluation is completed, by the engine.
 	// We want to prevent this from happening for the async store API calls we make while preserving tracing context.
 	ctx := tracing.CopyTraceContext(context.Background(), q.ctx)
-	ctx, cancel := context.WithTimeout(ctx, q.selectTimeout)
+	// The timeout value at q.ctx would contain least of {global Query timeout, per-request timeout}.
+	// For q.ctx, 'per-request timeout' is set pkg/api/query/v1.go (query(),query_range()) and later in call chain 'global Query timeout' is set at prometheus/promql/engine.go (exec()).
+	requestCancelTime, _ := q.ctx.Deadline()
+	requestTimeOut := time.Until(requestCancelTime)
+	ctx, cancel := context.WithTimeout(ctx, requestTimeOut)
```

Comment:
I think we don't need this. context we take timeout on is from `ctx := tracing.CopyTraceContext(context.Background(), q.ctx)` which already should have deadline. 

If we remove this code, your test fails?

Reply:
For q.ctx, 'per-request timeout' is set by [Thanos Query component ](https://github.com/thanos-io/thanos/blob/1205a222cb1e1be4862ef39b4c3a9c538056cf83/pkg/api/query/v1.go#L302)and later replaced by 'global Query timeout' at [downstream 'promql' package](https://github.com/prometheus/prometheus/blob/ad642a85c061443631404a72e7f1ecb4ab350b8e/promql/engine.go#L477) of upstream Prometheus.